### ISSUE TITLE
Distributed minio node process goes down if its logical drive is 100% written with objects data.

### ISSUE THREAD

**Issue by mkpavilion** [minio/minio#12028]
After successful distributed minio setup, keep uploading objects to it in order to fill up any end logical drive storage space 100%. After that the minio node process which owns that logical drive starts existing. It keep exiting even after re-started multiple times.
In my case, DMinio cluster is running over docker containers.

## Expected Behavior

If a logical drive within a Erasure Set has storage full then that Erasure Set shall stop taking any objects for writing. And just provide objects read support. In case the objects are deleted residing over that EC-set then it shall resume taking in new objects due to freed up storage space.

## Current Behavior

Distributed minio node process goes down if its logical drive is 100% written with objects data.

## Possible Solution

Minio node within a EC-set when finds out that its the logical drive is about to get 100% full then it shall communicate with other nodes too within that EC-set asking stop to take any new objects for writing under the set.
Minio nodes within the cluster keep syncing up intermittently exchanging status info. Leveraging this above msg communication can be fulfilled. In such case if there are more than one EC-set available within the Object STore namespace then any new objects writes shall go over them ignoring this filled up EC-Set.
Vice versa, when storage is freed up under that full EC-set, it shall broadcast again its status saying it can now consume newly uploaded objects under it.

## Steps to Reproduce (for bugs)

1. Prepare a DMinio setup with smallest EC-set (i used 16 size but i believe with 4 as well this issue shall get reproduced).
2. Use smallest possible minio logical drives so it shall take small time to repro the issue.
3. Keep uploaded objects to this EC until one logical drives gets full.
4. Minio node process which owns that drive shall start existing. If made service then it shall keep existing continuously.

## Regression

Not sure if this is regression as we tested it first time.

## Your Environment

* Version used (`minio --version`): minio version RELEASE.2021-03-26T00-00-41Z
* Server setup and configuration: minio nodes are running over independent docker containers. below is uname o/p from 
*  then docker container.
* Operating System and version (`uname -a`):Linux pds3-node1 5.4.19 #1 SMP -svn9789 Thu Apr 8 01:59:56 UTC 2021 x86_64 x86_64 x86_64 GNU/Linux

**Comment by klauspost** [comment]
What do you mean by "Minio node process [...] start existing"? That is not a term I know.

Your issue is rather dense to understand. Is it possible to rephrase it in simpler terms and maybe add the exact error messages that are logged/returned?

**Comment by mkpavilion** [comment]
By "minio node process" i meant "minio server endpoint within that cluster".
That process given below error and exits.

[pvld_controller_15   ] 2021-04-04 02:42:12,858 pds3-node15 root[54]: API: SYSTEM()
[pvld_controller_15   ] 2021-04-04 02:42:12,858 pds3-node15 root[54]: Time: 09:42:12 UTC 04/04/2021
[pvld_controller_15   ] 2021-04-04 02:42:12,858 pds3-node15 root[54]: DeploymentID: 94cc2a75-cf1a-44c6-acc9-9ad3765ef22c
[pvld_controller_15   ] 2021-04-04 02:42:12,858 pds3-node15 root[54]: Error: Storage resources are insufficient for the write operation .minio.sys/tmp/54114d49-fa90-48b7-a85f-1092df9aec48/10223c9b-f162-43b9-a4bf-1f4e9cf4f5c6/part.1 (cmd.InsufficientWriteQuorum)
[pvld_controller_15   ] 2021-04-04 02:42:12,858 pds3-node15 root[54]:        2: cmd/metacache-bucket.go:184:cmd.(*bucketMetacache).save()
[pvld_controller_15   ] 2021-04-04 02:42:12,858 pds3-node15 root[54]:        1: cmd/metacache-manager.go:83:cmd.(*metacacheManager).initManager.func1()
[pvld_controller_15   ] 2021-04-04 02:42:12,858 pds3-node15 root[54]:
[pvld_controller_15   ] 2021-04-04 02:42:12,858 pds3-node15 root[54]: API: SYSTEM()
[pvld_controller_15   ] 2021-04-04 02:42:12,858 pds3-node15 root[54]: Time: 09:42:12 UTC 04/04/2021
[pvld_controller_15   ] 2021-04-04 02:42:12,858 pds3-node15 root[54]: DeploymentID: 94cc2a75-cf1a-44c6-acc9-9ad3765ef22c
[pvld_controller_15   ] 2021-04-04 02:42:12,858 pds3-node15 root[54]: Error: Storage resources are insufficient for the write operation .minio.sys/tmp/54114d49-fa90-48b7-a85f-1092df9aec48/10223c9b-f162-43b9-a4bf-1f4e9cf4f5c6/part.1 (cmd.InsufficientWriteQuorum)

**Comment by klauspost** [comment]
@mkpavilion These are just errors and the server doesn't quit.

MinIO will stop accepting writes a while before the disks are filled, but temporary data to satisfy requests are still written.

You should have active monitoring of your disk space/inodes and take appropriate actions before they fill up. Once your disks are full, there is little we can do.

**Comment by klauspost** [comment]
Running this close to the limit will always cause problems.

**Comment by mkpavilion** [comment]
@klauspost On docker container based MinIO we found that the server quit when disk/s goes 100%. It is reproducible always.
Yes, we are certainly monitoring the disk space & inodes as well. And we raise appropriate alerts to user when consumption goes beyond high water mark level.

>> MinIO will stop accepting writes a while before the disks are filled, but temporary data to satisfy requests are still written.

I see. Then probably this could be a problem with docker based run (not sure0 as the server process is quitting definitely every time. But again from MinIO server perspective it should not matter OS environment and only left disk space shall matter here.

**Comment by klauspost** [comment]
@mkpavilion Maybe you should lower the threshold for notifications.

This is expected behavior. Since you are using our product commercially you might be interested in our subscription service where we can go into more details about your issue: https://min.io/pricing
